src/poly/evaluator.rs

```diff
@@ -0,0 +1,542 @@
+/// an evaluator will only be used to evaluate [`Ast`]s containing [`AstLeaf`]s obtained
+/// from itself. It should be set to the empty closure `|| {}`, because anonymous closures
+/// all have unique types.
+pub(crate) fn new_evaluator<E: Fn() + Clone, F: Field, B: Basis>(context: E) -> Evaluator<E, F, B> {
+    Evaluator {
+        polys: vec![],
+        _context: context,
+    }
+}
+
+impl<E, F: Field, B: Basis> Evaluator<E, F, B> {
+    /// Registers the given polynomial for use in this evaluation context.
+    ///
+    /// This API treats each registered polynomial as unique, even if the same polynomial
+    /// is added multiple times.
+    pub(crate) fn register_poly(&mut self, poly: Polynomial<F, B>) -> AstLeaf<E, B> {
+        let index = self.polys.len();
+        self.polys.push(poly);
+
+        AstLeaf {
+            index,
+            rotation: Rotation::cur(),
+            _evaluator: PhantomData::default(),
+        }
+    }
+
+    /// Evaluates the given polynomial operation against this context.
+    pub(crate) fn evaluate(
+        &self,
+        ast: &Ast<E, F, B>,
+        domain: &EvaluationDomain<F>,
+    ) -> Polynomial<F, B>
+    where
+        E: Copy + Send + Sync,
+        F: FieldExt,
+        B: BasisOps,
+    {
+        // Traverse `ast` to collect the used leaves.
+        fn collect_rotations<E: Copy, F: Field, B: Basis>(
+            ast: &Ast<E, F, B>,
+        ) -> HashSet<AstLeaf<E, B>> {
+            match ast {
+                Ast::Poly(leaf) => vec![*leaf].into_iter().collect(),
+                Ast::Add(a, b) | Ast::Mul(AstMul(a, b)) => {
+                    let lhs = collect_rotations(a);
+                    let rhs = collect_rotations(b);
+                    lhs.union(&rhs).cloned().collect()
+                }
+                Ast::Scale(a, _) => collect_rotations(a),
+                Ast::LinearTerm(_) | Ast::ConstantTerm(_) => HashSet::default(),
+            }
+        }
+        let leaves = collect_rotations(ast);
+
+        // Produce the rotated polynomials.
+        let rotated: HashMap<_, _> = leaves
+            .iter()
+            .cloned()
+            .map(|leaf| {
+                (
+                    leaf,
+                    B::rotate(domain, &self.polys[leaf.index], leaf.rotation),
+                )
+            })
+            .collect();
+
+        // We're working in a single basis, so all polynomials are the same length.
+        let poly_len = self.polys.first().unwrap().len();
+        // Check the level of parallelization we have available.
+        let num_threads = multicore::current_num_threads();
+        // We scale the number of chunks by a constant factor, to ensure that if not all
+        // threads are available, we can achieve more uniform throughput and don't end up
+        // waiting on a couple of threads to process the last chunks.
+        let num_chunks = num_threads * 4;
```

Comment:
This probably needs to be larger. The aim is to balance the overhead of having fewer-than-available threads executing the final few chunks, against the overhead of work stealing. The latter will be relatively small per chunk, and my intuition is that it's likely to still be smaller than a few percent until about `num_threads * 32`. We can estimate that the former is less than 1/32 at that point, i.e. also a few percent.

Reply:
I picked this as a starting point to build the logic around. Once the rest of the PR is complete, we can tune this in a production environment (i.e. Orchard) and see what kinds of overheads we are getting.

Reply:
I tried bumping the factor from `4` to `32`, and proving times for Orchard get slightly worse:
```
proving/bundle/1        time:   [2.6756 s 2.6925 s 2.7089 s]                              
                        change: [+0.3332% +1.0938% +1.8778%] (p = 0.02 < 0.05)
                        Change within noise threshold.
proving/bundle/2        time:   [2.6923 s 2.7061 s 2.7188 s]                              
                        change: [+1.4111% +2.1757% +2.9105%] (p = 0.00 < 0.05)
                        Performance has regressed.
proving/bundle/3        time:   [3.8655 s 3.8811 s 3.8968 s]                              
                        change: [+0.9132% +1.6442% +2.3634%] (p = 0.00 < 0.05)
                        Change within noise threshold.
proving/bundle/4        time:   [5.1000 s 5.1249 s 5.1464 s]                              
                        change: [+0.8925% +1.7243% +2.4899%] (p = 0.00 < 0.05)
```

Reply:
Right; the trade-off here is adding some average-case overhead in order to reduce the worst case when threads are contended. It's also worth trying 8 or 16.